3p/iframe-transport-client.js

```diff
@@ -33,61 +34,124 @@ export class IframeTransportClient {
+           */
+          (eventData['events']);
           user().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              user().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              const context = this.contextFor_(event.creativeId);
+              context.dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+              e.message);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    const vendor = tryParseJson(this.win_.name)['type'];
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, vendor));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    this.win_ = win;
+    this.iframeMessagingClient_ = iframeMessagingClient;
+    this.creativeId_ = creativeId;
+    this.vendor_ = vendor;
+
+    /** @private {?function(string)} */
+    this.listener_ = null;
+
+    user().assert(this.win_.onNewAmpAnalyticsInstance,
+        'Must implement onNewAmpAnalyticsInstance in ' +
+        this.win_.location.href);
+    this.win_.onNewAmpAnalyticsInstance(this);
+  }
+
+  /**
    * Registers a callback function to be called when an AMP analytics event
    * is received.
    * Note that calling this a second time will result in the first listener
    * being removed - the events will not be sent to both callbacks.
-   * @param {function(string,string)} callback
+   * @param {function(string)} listener
    */
-  onAnalyticsEvent(callback) {
-    this.listener_ = callback;
+  onAnalyticsEvent(listener) {
```

Comment:
Does anything need to be done to ensure this function name is properly externed (same for sendToCreative)?  Meaning it is not renamed as part of compilation?  Perhaps you already handled this... just curious

Reply:
Looking into this...

Reply:
Ok, made changes to amp.extern.js